src/v/storage/key_offset_map.cc

```diff
@@ -49,4 +49,209 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+        not_inserted_full,
+    };
+
+    // handle a non-normalized probe position
+    // returns true if key is inserted
+    auto handle_entry = [&](size_t index) {
+        ++probe_count_;
+        index = index % entries_.size();
+        auto& entry = entries_[index];
+        if (entry.empty()) {
+            if (full) {
+                return handle::not_inserted_full;
+            }
+            entry.digest = hash;
+            entry.offset = offset;
+            ++size_;
+            max_offset_ = std::max(max_offset_, offset);
+            return handle::inserted;
+        } else if (entry.digest == hash) {
+            if (offset > entry.offset) {
+                entry.offset = offset;
+                max_offset_ = std::max(max_offset_, offset);
+            }
+            return handle::inserted;
+        }
+        return handle::not_inserted;
+    };
+
+    ++search_count_;
+
+    // probe using the hash
+    probe probe(hash);
+    auto index = probe.next();
+    while (index.has_value()) {
+        const auto res = handle_entry(index.value());
+        if (res == handle::inserted) {
+            return seastar::make_ready_future<bool>(true);
+        }
+        if (res == handle::not_inserted_full) {
+            return seastar::make_ready_future<bool>(false);
+        }
+        const auto next_index = probe.next();
+        if (next_index.has_value()) {
+            index = next_index;
+        } else {
+            break;
+        }
+    }
+
+    // fall back to linear probe
+    const auto linear_base = index.value_or(0);
+    for (size_t probe = 0; probe < entries_.size(); ++probe) {
+        const auto res = handle_entry(linear_base + probe);
+        if (res == handle::inserted) {
+            return seastar::make_ready_future<bool>(true);
+        }
+        if (res == handle::not_inserted_full) {
+            return seastar::make_ready_future<bool>(false);
+        }
+    }
+
+    return seastar::make_ready_future<bool>(false);
+}
+
+model::offset hash_key_offset_map::max_offset() const { return max_offset_; }
+
+size_t hash_key_offset_map::size() const { return size_; }
+size_t hash_key_offset_map::capacity() const { return capacity_; }
+
+seastar::future<> hash_key_offset_map::reset(size_t size) {
+    co_await fragmented_vector_clear_async(entries_);
+    while (entries_.memory_size() < size) {
+        for (size_t i = 0; i < entries_.elements_per_fragment(); ++i) {
+            entries_.push_back(entry{});
+        }
+        if (seastar::need_preempt()) {
+            co_await seastar::maybe_yield();
+        }
+    }
+    size_ = 0;
+    max_offset_ = model::offset{};
+    if (entries_.size() > 0) {
+        capacity_ = std::max(
+          size_t(1),
+          static_cast<size_t>(
+            static_cast<double>(entries_.size()) * max_load_factor));
+    } else {
+        capacity_ = 0;
+    }
+    search_count_ = 0;
+    probe_count_ = 0;
+}
+
+seastar::future<> hash_key_offset_map::initialize() {
```

Comment:
nit: maybe consider naming this `reset()` or something? As is it seems like something we should always call before using the map, but I don't think that's the case (presumably we just need to use `reset(size_bytes)`?

Reply:
`reset(size)` implies `initialize()`. but in practice you really only want to call `reset(size)` once at boot-up, and then call `initialize()` each time you want to use the table to do a scan because it avoids freeing and reallocating all the memory.